cmd/maya-apiserver/spc-watcher/select_disk.go

```diff
@@ -0,0 +1,272 @@
+
+	err, usedDiskMap := k.getUsedDiskMap()
+	if err != nil {
+		return err, nil, pendingAllotment
+	}
+	err, usedNodeMap := k.getUsedNodeMap(spc)
+	if err != nil {
+		return err, nil, pendingAllotment
+	}
+	// nodeDiskMap is the data structure holding host name as key
+	// and nodeDisk struct as value
+	nodeDiskMap := make(map[string]*nodeDisk)
+	for _, value := range listDisk.Items {
+
+		// If the disk is already being used, do not consider this as a part for provisioning pool
+		if usedDiskMap[value.Name] == 1 {
+			continue
+		}
+		if usedNodeMap[value.Labels[string(v1alpha1.HostNameCPK)]] == 1 {
+			continue
+		}
+		// if no more allotment is required, stop processing
+		if pendingAllotment == 0 {
+			glog.Info("required pool allotment done")
+			break
+		}
+		if nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]] == nil {
+			// Entry to this block means first time the hostname will be mapped for the first time.
+			// Obviously, this entry of hostname(node) is for a usable disk and initialize diskCount to 1.
+			nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]] = &nodeDisk{diskList: []string{value.Name}, diskCount: 1}
+			// If pool type is striped the node qualifies for pool creation hence pendingAllotment decremented.
+			if poolType == string(v1alpha1.PoolTypeStripedCPK) {
+				pendingAllotment--
+			}
+		} else {
+			// Entry to this block means the hostname was already mapped and it has more than one disk and at least two disks.
+			nodeDisk := nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]]
+			// Increment the disk count
+			nodeDisk.diskCount++
+			// Add the current disk to the diskList for this node.
+			nodeDisk.diskList = append(nodeDisk.diskList, value.Name)
+			// If pool type is mirrored the node qualifies for pool creation hence pendingAllotment decremented.
+			if poolType == string(v1alpha1.PoolTypeMirroredCPK) {
+				if nodeDisk.diskCount == int(v1alpha1.MirroredDiskCountCPK) {
+					pendingAllotment--
+				}
+			}
+		}
+
+	}
+	return nil, nodeDiskMap, pendingAllotment
+}
+
+// diskSelector is the function that will select the required number of disks from qualified nodes
+// so as to provision storagepool
+
+func diskSelector(nodeDiskMap map[string]*nodeDisk, poolType string) []string {
+
+	// selectedDisk will hold a list of disk that will be used to provision storage pool, after a
+	// minimum number of node qualifies
+	var selectedDisk []string
+
+	// requiredDiskCount will hold the required number of disk that should be selcted from a qualified
+	// node for specific pool type
+	var requiredDiskCount int
+
+	// Range over the nodeDiskMap map to get the list of disks
+	for _, val := range nodeDiskMap {
+		// If pool type is striped, 1 disk should be selected
+		if poolType == string(v1alpha1.PoolTypeStripedCPK) {
+			requiredDiskCount = int(v1alpha1.StripedDiskCountCPK)
```

Comment:
requiredDiskCount assignment need not happen inside the loop. This is a one-time assignment and can be done above the loop.

Reply:
fixed